### ISSUE TITLE
Some questions about the code.

### ISSUE THREAD

**Issue by TOM-tym** [yaoyao-liu/class-incremental-learning#15]
Hi, thanks for your work.
There are some codes in [mnemonics-training/1_train/trainer/mnemonics.py](https://github.com/yaoyao-liu/class-incremental-learning/blob/main/mnemonics-training/1_train/trainer/mnemonics.py) I am confused about.

1. what's the different between two branch?
```python
                if iteration > start_iter:
                    tg_model = incremental_train_and_eval(self.args.epochs, tg_model, ref_model, free_model, ref_free_model, tg_optimizer, tg_lr_scheduler, trainloader, testloader, iteration, start_iter, cur_lamda, self.args.dist, self.args.K, self.args.lw_mr)
                else:
                    tg_model = incremental_train_and_eval(self.args.epochs, tg_model, ref_model, free_model, ref_free_model, tg_optimizer, tg_lr_scheduler, trainloader, testloader, iteration, start_iter, cur_lamda, self.args.dist, self.args.K, self.args.lw_mr)
```
I can't find and differences between two brances `if` and `else`.
original code: see [here](https://github.com/yaoyao-liu/class-incremental-learning/blob/main/mnemonics-training/1_train/trainer/mnemonics.py#L234-L237).

2. What does the function [process_fp](https://github.com/yaoyao-liu/class-incremental-learning/blob/main/mnemonics-training/1_train/utils/process_fp.py#L5-L23) do? As far as I am concerned, this function is just pass input in to the model, functioning the same with pytorch method forward().
3. What's the function of variable [tg_feature_model](https://github.com/yaoyao-liu/class-incremental-learning/blob/main/mnemonics-training/1_train/trainer/mnemonics.py#L172)? Is it just a alias of `tg_model` w/o the final fc layer?

It would be so kind of you if you could help me answer these questions.
Thanks again.

**Comment by yaoyao-liu** [comment]
Thanks for your interest in our work.

For Q1: Nothing different. I forget to merge these two lines while cleaning the code.
For Q2: The original version is [here](https://github.com/yaoyao-liu/class-incremental-learning/blob/main/adaptive-aggregation-networks/utils/process_fp.py). In mnemonics training, this function is just let the image pass the model.
For Q3: Yes.

**Comment by TOM-tym** [comment]
Hi, Thanks for your clarification. It helps a lot.

Pardon me for raising another question.
I assume that [line 325-373](https://github.com/yaoyao-liu/class-incremental-learning/blob/main/mnemonics-training/1_train/trainer/mnemonics.py#L325-L373) in `mnemonics.py` is the 'meta-level' optimazaion, corresponding to formula (7a), (7b), (8), (9), (10a), (10b) in the paper. Is that correct?
However, In my opinion, line 325-373 is just doing things below.

1. Make example set trainable. [line 325-326](https://github.com/yaoyao-liu/class-incremental-learning/blob/main/mnemonics-training/1_train/trainer/mnemonics.py#L325-L326)
2. Pass through image batch (sample from dataset $D_i$) into the model and get `q_feature`. [line 343-346](https://github.com/yaoyao-liu/class-incremental-learning/blob/main/mnemonics-training/1_train/trainer/mnemonics.py#L343-L346)
3. Pass through sample image batch to the model and get `all_cls_means`.
4. Calculate cross entropy loss.
In conclusion, turn example set trainable and apply it on the cross entropy loss. I don't find any code corresponding to formula (7b), i.e, train $\theta_i$ on $E_i$.
Please enlighten me about this.

Also, I can't find any code about “fine-tuning models on only exemplars” as shown in Line 18 in algorithm 1 in the paper.

**Comment by yaoyao-liu** [comment]
The code for “fine-tuning models on only exemplars” is here:
https://github.com/yaoyao-liu/class-incremental-learning/blob/6e5e46cd2f942b3f5eb338ff366c1c7e91f0e805/mnemonics-training/1_train/utils/compute_accuracy.py#L25-L45

**Comment by TOM-tym** [comment]
Thanks for your early reply. 
I am sorry for not reading your code completely. Thanks for pointing out.

What about another question?
>    I don't find any code corresponding to formula (7b), i.e, train $\theta_i$ on $E_i$.

Sorry to bother you again. Thanks.

**Comment by yaoyao-liu** [comment]
This part is related to Lines 349-369 in mnemonics.py. 
To speed up the training, the inner loop is replaced by a prototype classifier in the current version.

**Comment by TOM-tym** [comment]
Thanks for your explanation.
I am closing this issue now.
Best regards.

**Comment by TOM-tym** [comment]
Thanks for your explanation.
I am closing this issue now.
Best regards.
